### PR TITLE
Merge branch 'master' into rtd_sync_master_kip500_Dec04_2020

### PR DESCRIPTION
**Pull request by rondagostino**
Merge branch 'master' into rtd_sync_master_kip500_Dec04_2020

Conflicts:

    clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java
    clients/src/main/java/org/apache/kafka/common/protocol/Errors.java
    clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java
    clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java
    core/src/main/scala/kafka/Kafka.scala
    core/src/main/scala/kafka/cluster/Partition.scala
    core/src/main/scala/kafka/network/SocketServer.scala
    core/src/main/scala/kafka/server/AbstractBrokerToControllerChannelManager.scala
    core/src/main/scala/kafka/server/KafkaApis.scala
    core/src/main/scala/kafka/server/KafkaConfig.scala
    core/src/main/scala/kafka/server/LegacyBroker.scala
    core/src/main/scala/kafka/server/ReplicaManager.scala
    core/src/main/scala/kafka/tools/DumpLogSegments.scala
    core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
    core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
    core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala
    core/src/test/scala/unit/kafka/utils/TestUtils.scala

File Modified:

    clients/src/main/resources/common/message/BrokerRegistration{Request,Response}.json (change apiKey)
    core/src/main/scala/kafka/server/Kip500Broker.scala (have both alterIsrChannelManager and forwardingChannelManager as per LegacyBroker) 
    core/src/main/scala/kafka/server/Kip500BrokerToControllerChannelManager.scala (add channelName parameter)
    core/src/main/scala/kafka/server/LegacyBrokerToControllerChannelManager.scala (add channelName parameter)
    core/src/main/scala/kafka/server/MetadataCache.scala (UUID => Uuid)
    core/src/test/scala/unit/kafka/server/metadata/PartitionMetadataProcessorTest.scala (UUID => Uuid)
    core/src/test/scala/unit/kafka/tools/DumpLogSegmentsTest.scala (UUID => Uuid)
    metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java (UUID => Uuid)

### CODE REVIEW

**Comment by rondagostino** [comment]
There are 2 failing tests in `KafkaApisTest` related to envelope processing.  They are almost certainly testing issues rather than bugs since we've had failing tests in this file before.

1. `testEnvelopeRequestHandlingAsController()` (NPE)
2. `testEnvelopeRequestWithNotFromPrivilegedListener()` (had to temporarily code `assertTrue(false)` since `assertTrue(capturedResponse.getValue.isInstanceOf[CloseConnectionResponse])` does not compile due to changes in https://github.com/confluentinc/kafka/pull/441.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 1]
```diff
@@ -412,97 +420,640 @@ class KafkaApisTest {
     results.foreach(r => {
       assertEquals(Errors.NONE.code, r.errorCode)
       assertFalse("Should return configs", r.configs.isEmpty)
-      r.configs.forEach(c => {
+      r.configs.asScala.filter { c => c.name != KafkaConfig.EnableMetadataQuorumProp }.foreach(c => {
```
Comment (rondagostino) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:423]: @abbccdda `enable.metadata.quorum` has no documentation, so I had to exclude that explicitly from the test `KafkaApisTest.testDescribeConfigsWithDocumentationKip500` — is this something that needs to be fixed in AK?

**Comment by rondagostino** [comment]
There was a significant conflict between https://github.com/confluentinc/kafka/pull/421/ (`Add ControllerApis`) and https://github.com/apache/kafka/pull/9103/ (`KAFKA-10181: Use Envelope RPC to do redirection for (Incremental)AlterConfig, AlterClientQuota and CreateTopics`). @mumrah @abbccdda Please review what I did to resolve the issues, as describe here:

In `KafkaApis.scala`: **REPLACE** invocations of `closeConnection(request, errorCounts)`in the AK code with `requestChannel.closeConnection()` (the `Add ControllerApis` PR has moved it).

In `KafkaApis.scala`: **REPLACE** invocations of `sendErrorResponseMaybeThrottle(request, throwable)` in the AK code with `apisUtils.sendErrorResponseMaybeThrottle()` (the `Add ControllerApis` PR has moved it).
In `ApisUtils.scala` **UPDATE** definition of `sendErrorResponseMaybeThrottle(request, throwable)` to reflect updated KAFKA-10181 definition in `KafkaApis` (the `Add ControllerApis` PR has moved it).

In `KafkaApis.scala`: **REPLACE** invocations of `sendResponseMaybeThrottle(request, createResponse, onComplete)` in the AK code with `apisUtils.sendResponseMaybeThrottle(request, createResponse)` (the `Add ControllerApis` PR has moved it).
In `ApisUtils.scala` **UPDATE** definition of `sendResponseMaybeThrottle(request, createResponse, onComplete)` to reflect updated KAFKA-10181 definition in `KafkaApis` (the `Add ControllerApis` PR has moved it).

In `ApisUtils.scala` **DELETE** `def sendResponseMaybeThrottle(controllerMutationQuota, request, createResponse, onComplete)` as it is no longer used now that we have `private def sendResponseMaybeThrottleWithControllerQuota()` in `KafkaApis.scala` (KAFKA-10181 introduced that method, and I left it there instead of moving it into `ApisUtils.scala`).

In `KafkaApis.scala`: **REPLACE** invocations of `authorize()` in the AK code with `apisUtils.authorize()` (the `Add ControllerApis` PR has moved it).

In `KafkaApis.scala`: **REPLACE** invocations of `sendResponse(request, responseOpt, onComplete)` in the AK code with `requestChannel.sendResponse()` (the `Add ControllerApis` PR has moved it).
In `RequestChannel.scala` **UPDATE** definition of `sendResponse(request, responseOpt, onComplete)` to reflect updated KAFKA-10181 definition in `KafkaApis` (the `Add ControllerApis` PR has moved it).

**Comment by rondagostino** [comment]
@aloknnikhil Regarding the 2 failing tests in `KafkaApisTest` related to envelope processing, I will dig in and try to figure out how to fix them, but I may need your help.  Will let you know.

**Comment by cmccabe** [comment]
LGTM

**Pull request merged**